Memory efficient single core mode

chkngit opened the pull request:
## Description

- Use the least possible resource allocation  to have an efficient single core operating mode.
- Fix deadlock with -lad 0
- Latency(controlled by number of parent pcs = number of input) went down from 72 frames to 24 frames:
24 = 17(GOP) + 6(SCD/ALTREF) +1(EOS) + 0 (Needed Pictures for Look-Ahead)
Ref = 17 assuming a 5L GOP.  
Pa-Ref = 25  + 6(SCD/ALTREF) +1(EOS) ,  assuming a 5L GOP. 

- Change output bistream allocation to be done at run time to serve a single threaded app.
- Changes have been made to picture decision to remove the 6-frames delay in case  Scene-Change and ALTREF are OFF.
- Added the ability to unpin the single core execution to allow multiple single core encodes on the same machine without them being pinned to all core 0 [execution is unpinned by default]

## Performance
- For a 5L GOP,  1080p single core mode (fixed QP). Memory allocation went down from 4405 MB to 2032 MB.
- 100% memory reduction with -lad 0 and -lp 1

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
     sequence_control_set_ptr->overlay_input_picture_buffer_init_count   = sequence_control_set_ptr->static_config.enable_overlays ?
                                                                           (2 << sequence_control_set_ptr->static_config.hierarchical_levels) + SCD_LAD : 1;
+#if SERIAL_MODE
+    //Future window in SCD/TemporalFiltering
```
How about "Future frames window in SCD/TemporalFiltering" instead?

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
+
+    if (core_count == SINGLE_CORE_COUNT) {
+
+        //Look-Ahead. Picture-Decision outputs pictures by group of mini-gops so  the needed pictures for a certain lad should be rounded up to the next multiple of MiniGopSize.
```
Please use "the needed pictures for a certain look-ahead distance (LAD) should be" instead of "the needed pictures for a certain lad should be"

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
+        //Look-Ahead. Picture-Decision outputs pictures by group of mini-gops so  the needed pictures for a certain lad should be rounded up to the next multiple of MiniGopSize.
+        uint32_t mg_size = 1 << sequence_control_set_ptr->static_config.hierarchical_levels;
+        uint32_t needed_lad_pictures = ((sequence_control_set_ptr->static_config.look_ahead_distance + mg_size - 1) / mg_size) * mg_size;
+        //printf("Needing %i for lad\n", needed_lad_pictures);
```
Remove the printf statement.

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
+        uint32_t needed_lad_pictures = ((sequence_control_set_ptr->static_config.look_ahead_distance + mg_size - 1) / mg_size) * mg_size;
+        //printf("Needing %i for lad\n", needed_lad_pictures);
+
+        /*To accomodate FFMPEG EOS, 1 frame delay is needed in Ress-Coord.
```
Use "To accomodate FFMPEG EOS, 1 frame delay is needed in Resource coordination."

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
+        //printf("Needing %i for lad\n", needed_lad_pictures);
+
+        /*To accomodate FFMPEG EOS, 1 frame delay is needed in Ress-Coord.
+           note that we have the option to not add 1 frame delay of RessCoord.in this case we have wait for first I frame to be released back to be able to start first base(16).
```
Use "of Resource Coordination. In this case, we have to wait for the first I frame" instead of "of RessCoord.in this case we have wait for first I frame"

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
+        uint32_t eos_delay = 1;
+
+
+        //Minumum needed input pictures in the pipeline
```
Use "Minimum input pictures needed in the pipeline."

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
+            sequence_control_set_ptr->picture_control_set_pool_init_count =
+            ((mg_size + 1) + eos_delay + sequence_control_set_ptr->scd_delay) * 2 + //min to get a mini-gop in PD - release all and keep one.
+            needed_lad_pictures +
+            needed_lad_pictures / mg_size + 1;// number of overlays in the LAD
```
How about "Number of overlays in the look-ahead window"?

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
+        //Pic-Manager will inject one child at a time.
+        sequence_control_set_ptr->picture_control_set_pool_init_count_child = 1;
+
+        sequence_control_set_ptr->reference_picture_buffer_init_count = 17;     //min to sustain flow (RA-5L-MRP-ON) 7 pictures from previous MGs + 10 needed for curr MG
```
If ""MG" stands for mini-GoP, then use "mini-GoP" instead.

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
+        sequence_control_set_ptr->picture_control_set_pool_init_count_child = 1;
+
+        sequence_control_set_ptr->reference_picture_buffer_init_count = 17;     //min to sustain flow (RA-5L-MRP-ON) 7 pictures from previous MGs + 10 needed for curr MG
+        sequence_control_set_ptr->pa_reference_picture_buffer_init_count = 25 + //min to sustain flow (RA-5L-MRP-ON) -->TODO: derive numbers for other GOP Structures.
```
Same here as above.

fouedb commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -476,6 +476,64 @@ EbErrorType load_default_buffer_configuration_settings(
+    //printf("OUTPUT %i \n", sequence_control_set_ptr->output_stream_buffer_fifo_init_count);
+    //printf("REF  %i \n", sequence_control_set_ptr->reference_picture_buffer_init_count);
+    //printf("CHILD %i \n", sequence_control_set_ptr->picture_control_set_pool_init_count_child);
+    //printf("OVERLAY %i \n", sequence_control_set_ptr->overlay_input_picture_buffer_init_count);
```
Delete these lines.

fouedb commented on Source/Lib/Common/Codec/EbSequenceControlSet.h:
```diff
@@ -126,6 +126,10 @@ extern "C" {
         EbBool                                  enable_altrefs;
+
+#if SERIAL_MODE
+        uint32_t                                scd_delay; //number of frame delay needed to implement future window for algorithms like ScenChange or TemporalFiltering.
```
Use "Number of delay frames needed to implement future window for algorithms such as SceneChange or TemporalFiltering."

fouedb commented:
Start all comments with capital letters.

AlexBoswellVCD commented on Source/Lib/Common/Codec/EbDefinitions.h:
```diff
@@ -61,7 +68,7 @@ extern "C" {
 #define INTER_INTRA_CLASS_PRUNING    1
 
-#define FIX_ALTREF                   1 // Address ALTREF mismatch between rtime-m0-test and master: fixed actual_future_pics derivation, shut padding of the central frame, fixed end past frame index prior to window shrinking
+#define FIX_ALTREF                   0 // Address ALTREF mismatch between rtime-m0-test and master: fixed actual_future_pics derivation, shut padding of the central frame, fixed end past frame index prior to window shrinking
```
I found that only the lines here are responsible for the r2r (https://github.com/OpenVisualCloud/SVT-AV1/commit/a5dcbf2ecdf6bda7db01aa359fc07e9ad1255871#diff-051be3f82f0c4e17fd579aa4eda9ba62R2295-R2297)

chkngit commented:
@fouedb done the changes thank you.
@hassount 
- fixed the mismatch was due non initialized global motion data. Just got triggered by the new dynamics.
- put back the ALT-REF-FIX and disabled only the padding if statement.
- added a min to make sure -lad 0 works with the old buffer counts and without -lp 1

hassount commented:
> @fouedb done the changes thank you.
> @hassount
> 
> * fixed the mismatch was due non initialized global motion data. Just got triggered by the new dynamics.
> * put back the ALT-REF-FIX and disabled only the padding if statement.
> * added a min to make sure -lad 0 works with the old buffer counts and without -lp 1

@chkngit added the ability to unpin the single core execution to allow multiple single core encodes on the same machine without them being pinned to all core 0

hassount commented:
@joelsole @mafonsoN any issues with this ?
